EB-282: Add copy-to-clipboard button to facilitate copy of citations

RMCrean opened the pull request:
Hey, 

This PR is is for adding copy-to-clipboard buttons for the citations and any code blocks. This PR doesn't include embedding the version info directly into the citations but it should be fine to include that at a later stage (in the shortcode) once ready (dependent on PR #62 ). 

I followed this [guide which summarizes quite nicely the steps required](https://logfetch.com/hugo-add-copy-to-clipboard-button/). There are some differences but the general idea is still there. 

**Adding a copy-to-clipboard button :** 
1. For a citation, Wrap the citation in the hugo shortcode `citation_block.html `, see `hugo/content/citation/_index.md` for examples. 

![Screencast from 2024-11-14 15_05_46](https://github.com/user-attachments/assets/3f7fe4a8-a533-4885-93b0-21d3d550aff1)

2. For code blocks, use standard markdown format, i.e.:  ``` 
Code blocks already marked with this are automatically recognized and provided with a copy-to-clipboard button. 
See examples of what this looks like on: `http://localhost:8080/contribute/supported_file_formats/`

**Notes:** 
1. Initially I tried using a shortcode to create the code blocks (as done for the citations). This caused the markdown linter to go crazy and seen as how it didn't provide any extra functionality/benefit I think it is best to stick to using: ``` for code blocks. 
2. Using the shortcode `citation_block.html ` is a bit of a hack. It essentially puts the citation text inside the same HTML structure that a code block has. This enables it to be recognized by `hugo/layouts/partials/copy_to_clipboard.html` 
3. I put the copy-to-clipboard button outside the code/citation block. I read somewhere that this is good to avoid putting it inside the block as it can end up sitting on top of the code/text if it makes it all the way to end of the line. 
4. The styling of the code block and citations blocks are identical with the exception that wrapping lines is allowed in the citation blocks.
5. Updated Hugo version to take advantage of https://gohugo.io/functions/strings/trimspace/

RMCrean commented:
("Build hugo site and test site with Lighthouse CI " failure was because the Hugo version is too old in the workflow file). Lighthouse will be removed in PR #62 so no need to do anything about it I think).

apfuentes commented:
Hi @RMCrean, really nice job adding the copy-to-clipboard feature! I checked both, the citation and the code blocks and both work and look nice.

I am now wondering more about UX. Users perhaps may rush and don't realize that they have to replace the text within the placeholders (which by the way it is nicely mentioned in the text below each citation block, thanks @brinkdp). Thus, I wonder if it would be possible to highlight in red (or any other contrasting color) the parts that need to be replaced, e.g. "access date" or the "version number" below:

Swedish Reference Genome Portal (access date), SciLifeLab Data Centre, version (version number) from https://genomes.scilifelab.se/, [RRID:SCR_026008](https://rrid.site/data/record/nlx_144509-1/SCR_026008/resolver?q=rrid:scr_026008).

brinkdp reviewed (approved):
Very nice, thanks! 

Sounds like a good idea to separate the `{{< citation_block >}}` and ``` to keep us sane from the linter output 😄 

I like how the box makes the citation info pop on the cite us page. I wonder, for UX reasons, if we should consider using a different styling for the citation and code blocks. I am primarily thinking about the font: the `courier new`-like font in the code blocks signal code, but an APA citation is not necessarily considered code for most users (LaTEX users aside). 

I also wonder if I would prefer the copy button to be positioned underneath the box than above it? My thinking is that a right-aligned underside button aligns with the eye movement, reading left-to-right, from top to bottom. I don't know if there are UI/UX best practices for this? Do you know @apfuentes?

Do we also want to apply this to each species description tab? Were we saying that we would go for a .bib there instead?

brinkdp commented:
> I wonder if it would be possible to highlight in red (or any other contrasting color) the parts that need to be replaced, e.g. "access date" or the "version number" below:

This is good idea! Like Rory described in the first post, the plan is eventually to automate populating the placeholders with date and version. But in the meantime, this could be useful. But if the trunk PR is soon ready for merge, perhaps it is better to wait?

RMCrean commented:
(I am working on this branch now that #62 is merged), will come back when ready for a round of review :)

RMCrean commented:
Hi all, 

I've now updated the {{< citation_block >}} shortcode to replace any of the following (matching is case sensitive):
- YEAR 
- DATE_ACCESSED 
- VERSION_NUMBER

When these words are specified inside the `{{< citation_block >}} ` shortcode of a markdown file they will be replaced accordingly. 

A normal dev build of the portal (e.g.:)

```docker rm -f "genome-portal" ; ./scripts/dockerbuild hugo ; SWG_TAG=local ./scripts/dockerserve```

would show up with the version number set to "X.X.X":

To test how it would look like in a production build you can provide the environment variable `HUGO_GIT_REF_NAME` in the docker build command:

```docker rm -f "genome-portal" ; HUGO_GIT_REF_NAME="v1.0.0" ./scripts/dockerbuild hugo ; SWG_TAG=local ./scripts/dockerserve```

Here is a preview: 
![Screenshot from 2024-11-19 13-38-19](https://github.com/user-attachments/assets/402013a8-dce2-4664-a8e9-505c107badf4)
(species name is the one thing I haven't replaced)

@brinkdp , can you make sure I got the APA citation formatting correct please? - I think so at least. This especially includes how I formatted the dates, and my use of brackets around things like the year/date accessed. I also reworded some of the texts (to reflect changes in placeholders so it would be good to have that checked over). (Please feel free to directly commit any changes you see fit). 

Once we have this page all correctly setup/formatted, we could move onto updating the species page citations, which would perhaps be best as a separate PR?

kwentine commented on hugo/layouts/partials/copy_to_clipboard.html:
```diff
@@ -0,0 +1,4 @@
+{{ if (findRE "<pre" .Content 1) }}
```
For an alternative approach to detect copy-blocks, see [render hook examples](https://gohugo.io/render-hooks/code-blocks/#examples). The idea is to set `$.Page.Store.Set "needs_clipboard" true` in the shortcode template, and then use `{{ if $.Page.Store.Get "needs_clipboard" }}` Not 100% sure if what works for render hooks will work for shortcodes though.

kwentine commented on hugo/layouts/partials/copy_to_clipboard.html:
```diff
@@ -0,0 +1,4 @@
+{{ if (findRE "<pre" .Content 1) }}
+    {{ $clipboardJS := resources.Get "js/clipboard.js" }}
+    <script>{{ $clipboardJS.Content | safeJS }}</script>
```
Why inlining the script rather than `<script src="{{ $clipboardJS.RelPermalink }}"/>` ?

RMCrean replied:
Fair point, I've added this now and fingerprinted it to handle caching (not that we probably expect this file to update much). Commit 57e7b0936a239158e156cbc775ace11e22898d31

kwentine commented on hugo/layouts/partials/copy_to_clipboard.html:
```diff
@@ -0,0 +1,4 @@
+{{ if (findRE "<pre" .Content 1) }}
```
For an alternative approach to detect copy-blocks, see [render hook examples](https://gohugo.io/render-hooks/code-blocks/#examples). The idea is to set `$.Page.Store.Set "needs_clipboard" true` in the shortcode template, and then use `{{ if $.Page.Store.Get "needs_clipboard" }}` Not 100% sure if what works for render hooks will work for shortcodes though.

RMCrean replied:
I've made some progress with render hooks now, I'll look into this bit now!

RMCrean replied:
Thanks for the link. Yes I can make this work and remove the need for the file copy_to_clipboard. A small downside was needing to add a new if block to `hugo/layouts/_default/baseof.html`, but I don't think that does much harm overall. Added this in commit 1bb8a7d3249a65dce7432f659ee8f3595d51d299

kwentine commented on hugo/assets/js/clipboard.js:
```diff
@@ -0,0 +1,56 @@
+// Adapted from https://logfetch.com/hugo-add-copy-to-clipboard-button/
+const svgCopy =
```
As a general remark, it may be worthwhile to try and reduce the amount of DOM manipulation in this snippet.
For example, you could include in the shortcode:
```
<button>
    <img src="assets/copy.svg">
</button>
<button>
    <img src="assets/paste.svg">
</button>
```

RMCrean replied:
Agreed, I've done that now in commit 802f3b94709a2f89749b2fc447e4150b5fb7508.

kwentine commented on hugo/layouts/shortcodes/citation_block.html:
I would encourage you to read the Hugo documentation section about [Render Hooks](https://gohugo.io/render-hooks/code-blocks/). In short, it provides a mechanism to override the way Hugo renders native Mardown syntax.

You could define `render-codeblock-citation.html` and then use:
```markdown
```citation
En herrgårdssägen, Selma Lagerlöf
\```
```
And if you also define `render-codeblock.html` it will be used by Hugo when triple quotes are used (without the `citation` info).
But maybe you tried this approach and that is what confused the linter ?

RMCrean replied:
Thanks for the tip, I've implemented this now in commit 802f3b94709a2f89749b2fc447e4150b5fb7508 which while it took a little while to understand/implement, made the code a lot better I think.

brinkdp commented on hugo/content/citation/_index.md:
```diff
@@ -25,31 +27,36 @@ You can also cite a specific species page from the Genome Portal. The easiest wa
 **In-text citation (APA format)**:\
-The [_Species name_] page in the Swedish Reference Genome Portal (access date).
+{{< citation_block >}}
+The <span class="highlight">[_Species name_]</span> page in the Swedish Reference Genome Portal (DATE_ACCESSED).
```
```suggestion
The <span class="highlight">[_Species name_]</span> page in the Swedish Reference Genome Portal (Retrieved DATE_ACCESSED).
```

brinkdp commented on hugo/content/citation/_index.md:
```diff
@@ -25,31 +27,36 @@ You can also cite a specific species page from the Genome Portal. The easiest wa
 
-"Visualisations of the genome assembly and annotation tracks from this publication can be found in the Swedish Reference Genome Portal (access date) at <https://genomes.scilifelab.se/[species_name]>."
+{{< citation_block >}}
+"Visualisations of the genome assembly and annotation tracks from this publication can be found in the Swedish Reference Genome Portal (DATE_ACCESSED) at <https://genomes.scilifelab.se/[species_name]>."
```
```suggestion
"Visualisations of the genome assembly and annotation tracks from this publication can be found in the Swedish Reference Genome Portal (Retrieved DATE_ACCESSED) at <https://genomes.scilifelab.se/[species_name]>."
```

RMCrean replied:
Didn't add this to batch the first time by accident, but have now committed it.

kwentine commented on hugo/layouts/shortcodes/citation_block.html:
```diff
@@ -0,0 +1,18 @@
+
+<!-- get the version number if prod build or set to X.X.X. if not -->
+{{ $version := partial "GetVersion.html" }}
+{{ $portalVersion := "X.X.X" }}
+{{ if $version.tag }}
+    {{ $portalVersion = index $version "tag" }}
+    {{ $portalVersion = replace $portalVersion "v" "" }}
+{{ end }}
+
```
🥷 🐿️ 
```suggestion
{{ $portalVersion := $version.tag | default "X.X.X" | strings.TrimPrefix "v" }}
```

kwentine replied:
Also, if the "v" is a hassle feel free to strip it in `GetVersion.html` (I think we don't use it for the JBrowse version in the footer)

RMCrean replied:
That was cool! I've add this in commit 3c2af9bc6cc2c83a5a87e87c4a6305c6db66ee5b (couldn't commit suggestion as this code has moved to a the render hooks file :) )

kwentine commented on hugo/layouts/shortcodes/citation_block.html:
```diff
@@ -0,0 +1,18 @@
+    {{ $portalVersion = replace $portalVersion "v" "" }}
+{{ end }}
+
+<!-- can't pipe the replace command, hence multiline-->
```
`strings.ReplaceRE` can be piped though, if you are tempted 🤷🏻‍♂️

RMCrean replied:
Haha, I put that comment because I thought you'd ask otherwise, good to know there is a way around it, will include it shortly!

kwentine replied:
You can put `<!-- noqå -->` 🤣

RMCrean replied:
```{{ $content = $content | strings.ReplaceRE "DATE_ACCESSED" $date -1 | strings.ReplaceRE "YEAR" $year -1 | strings.ReplaceRE "VERSION_NUMBER" $portalVersion -1 }}``` works indeed! 
But I will keep it multline as I think the line is too long otherwise

kwentine replied:
Problem solved:
```
{{ $content = $content | 
strings.ReplaceRE "DATE_ACCESSED" $date -1 | 
strings.ReplaceRE "YEAR" $year -1 | 
strings.ReplaceRE "VERSION_NUMBER" $portalVersion -1 
}}
```

RMCrean commented on hugo/content/citation/_index.md:
```diff
@@ -25,31 +27,36 @@ You can also cite a specific species page from the Genome Portal. The easiest wa
 
-"Visualisations of the genome assembly and annotation tracks from this publication can be found in the Swedish Reference Genome Portal (access date) at <https://genomes.scilifelab.se/[species_name]>."
+{{< citation_block >}}
+"Visualisations of the genome assembly and annotation tracks from this publication can be found in the Swedish Reference Genome Portal (DATE_ACCESSED) at <https://genomes.scilifelab.se/[species_name]>."
```
Note to self, should also remove the quotes now that it is in a block

RMCrean commented:
@brinkdp, thanks for going through the citations and checking them over, we can save your comments/suggestions comments till our meeting later today then. I think I am favor of including the word "Retrieved" for clarity, but we can see what others say. 

And agreed the grape color choice is strong, happy to edit/for it to be edited though.  

@kwentine, thanks for the suggestions. I think the render hooks approach is a lot cleaner. In the process I changed strategies a bit to. Now both the copy and copied buttons are present on the page already and then the JS just controls which one is visible. 

I now think I have addressed your comments, so please feel free to take a look when you have time and see if you agree. :) 

For fun I might implement a playwright test or two on these pages to validate the copy-to-clipboard functionality.

RMCrean commented on hugo/content/citation/_index.md:
```diff
@@ -25,31 +28,38 @@ You can also cite a specific species page from the Genome Portal. The easiest wa
 The general pattern for citing a species page can also be filled out using the following template:
 
 **In-text citation (APA format)**:\
```
```suggestion
**In-text citation (APA format)**:
```

RMCrean commented on hugo/content/citation/_index.md:
```diff
@@ -25,31 +28,38 @@ You can also cite a specific species page from the Genome Portal. The easiest wa
 
 The Genome Portal is operated by the <a target="_blank" href="https://scilifelab.se/data">SciLifeLab Data Centre</a> and partners. All of the source code used to generate the website is available on [GitHub](https://github.com/ScilifelabDataCentre/genome-portal/). The code is published under an MIT license. If you have reused the code or otherwise want to cite the code, please use the following:
 
 **APA format:**\
```
```suggestion
**APA format:**
```

RMCrean commented on hugo/content/citation/_index.md:
```diff
@@ -58,4 +68,7 @@ All datasets included in the Genome Portal are public, but the means as to how t
 You might also want to consider citing the JBrowse 2 genome browser, especially if you have used the genome browser to work with the data. The developers provide their own citation information on the [JBrowse website](https://jbrowse.org/). At the time of writing, the JBrowse developers ask users to cite the following publication:
 
 **APA format:**\
```
```suggestion
**APA format:**
```

kwentine commented on hugo/layouts/_default/_markup/render-codeblock.html:
```diff
@@ -0,0 +1,41 @@
+{{- $style := index .Attributes "style" | default "code" -}}
+
+<div class="clipboard">
+    <button id="copy-button" class="clipboard-button" type="button">
```
The HTML specification demands that `id` attribute values [be unique](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id) in the document. So maybe a class would be better.

RMCrean replied:
Good point, I've updated this in commit 2ea3400dccce2c949380bce104d2ad0aa6fcd730

kwentine reviewed (approved):
Great work! 👍🏻 And as a followup question: is counting clicks on the citation buttons something Matomo can do 🙃  ?

RMCrean commented on hugo/assets/js/clipboard.js:
```diff
@@ -0,0 +1,43 @@
+    );
+};
+
+// create the copy to clipboard button above the code/citation block
```
Good spot! Have removed the function in commit 2ea3400dccce2c949380bce104d2ad0aa6fcd730

RMCrean commented on hugo/assets/js/clipboard.js:
```diff
@@ -0,0 +1,43 @@
+};
+
+document.addEventListener("DOMContentLoaded", () => {
+    attachClipboardListeners(navigator.clipboard);
```
We'll say it was for that reason haha. I've updated the code to reflect it is already available in the global scope though commit 2ea3400dccce2c949380bce104d2ad0aa6fcd730 

Let's not speak of tests on a Friday surely ;)

RMCrean commented:
> Great work! 👍🏻 And as a followup question: is counting clicks on the citation buttons something Matomo can do 🙃 ?

Thanks Quentin, will merge in just a moment. Out of the box Matomo tells us the number of visits to a given page (and for how long) like the cite-us page but no such detail about specific button clicking events.
